# Lab 3

## Step 1

Question 1: How does early academic record impact college completion prediction? 

Question 2: Do academic background and specialization play a role in salary outcomes? 

## Step 2

##### College Completion Dataset

Independent Business Metric: grad_150_value

##### Data Preparation

In [ ]:
import kagglehub
import os
import pandas as pd
import numpy as np 
import matplotlib
from sklearn.model_selection import train_test_split 
from sklearn.preprocessing import MinMaxScaler, StandardScaler

path = kagglehub.dataset_download("thedevastator/boost-student-success-with-college-completion-da")

dataset_path = "/home/vscode/.cache/kagglehub/datasets/thedevastator/boost-student-success-with-college-completion-da/versions/1"
print(os.listdir(dataset_path))

['cc_state_sector_details.csv', 'cc_state_sector_grads.csv', 'cc_institution_grads.csv', 'cc_institution_details.csv']


In [86]:
file_name = "cc_institution_details.csv" 
df = pd.read_csv(os.path.join(dataset_path, file_name))

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3798 entries, 0 to 3797
Data columns (total 63 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   index                                 3798 non-null   int64  
 1   unitid                                3798 non-null   int64  
 2   chronname                             3798 non-null   object 
 3   city                                  3798 non-null   object 
 4   state                                 3798 non-null   object 
 5   level                                 3798 non-null   object 
 6   control                               3798 non-null   object 
 7   basic                                 3798 non-null   object 
 8   hbcu                                  94 non-null     object 
 9   flagship                              50 non-null     object 
 10  long_x                                3798 non-null   float64
 11  lat_y            

In [87]:
df2 = df[['level','control','student_count','awards_per_value','awards_per_state_value','awards_per_natl_value',
          'exp_award_value', 'exp_award_state_value', 'exp_award_natl_value', 'ft_pct','med_sat_value',
          'aid_value','grad_150_value']]

In [88]:
df2 = df2.dropna()

In [89]:
df2[['level','control']] = df2[['level','control']].astype('category')

print(df2.dtypes)

level                     category
control                   category
student_count                int64
awards_per_value           float64
awards_per_state_value     float64
awards_per_natl_value      float64
exp_award_value              int64
exp_award_state_value        int64
exp_award_natl_value         int64
ft_pct                     float64
med_sat_value              float64
aid_value                  float64
grad_150_value             float64
dtype: object


In [90]:
abc = list(df2.select_dtypes('number'))

df2[abc] = MinMaxScaler().fit_transform(df2[abc])
df2

level                 control  student_count  awards_per_value  \
0     4-year                  Public       0.077136          0.212062   
1     4-year                  Public       0.222567          0.342412   
3     4-year                  Public       0.109244          0.342412   
4     4-year                  Public       0.102608          0.161479   
5     4-year                  Public       0.572686          0.291829   
...      ...                     ...            ...               ...   
3572  4-year  Private not-for-profit       0.015751          0.262646   
3639  4-year                  Public       0.076629          0.328794   
3651  4-year  Private not-for-profit       0.000371          0.291829   
3717  4-year  Private not-for-profit       0.008139          0.184825   
3794  4-year  Private not-for-profit       0.000800          0.258755   

      awards_per_state_value  awards_per_natl_value  exp_award_value  \
0                   0.172680               0.531915         0.175872   
1                   0.172680               0.531915         0.227992   
3                   0.172680               0.531915         0.107559   
4                   0.172680               0.531915         0.221081   
5                   0.172680               0.531915         0.125812   
...                      ...                    ...              ...   
3572                0.311856               0.638298         0.248117   
3639                0.304124               0.531915         0.077117   
3651                0.226804               0.638298         0.287266   
3717                0.275773               0.638298         0.308513   
3794                0.265464               0.638298         0.176930   

      exp_award_state_value  exp_award_natl_value    ft_pct  med_sat_value  \
0                  0.311746              0.474983  0.922597       0.180876   
1                  0.311746              0.474983  0.659176       0.552995   
3                  0.311746              0.474983  0.680400       0.592166   
4                  0.311746              0.474983  0.887640       0.188940   
5                  0.311746              0.474983  0.877653       0.581797   
...                     ...                   ...       ...            ...   
3572               0.288542              1.000000  0.977528       0.465438   
3639               0.062524              0.474983  0.575531       0.465438   
3651               0.500767              1.000000  0.915106       0.465438   
3717               0.510470              1.000000  0.943820       0.569124   
3794               0.616908              1.000000  0.901373       0.464286   

      aid_value  grad_150_value  
0      0.147490        0.295132  
1      0.121398        0.542596  
3      0.135236        0.490872  
4      0.150312        0.255578  
5      0.227894        0.676471  
...         ...             ...  
3572   0.353476        0.585193  
3639   0.094589        0.433063  
3651   0.093425        0.000000  
3717   0.458313        0.622718  
3794   0.203263        0.522312  

[1319 rows x 13 columns]

In [91]:
category_list = list(df2.select_dtypes('category')) 
category_list
df3 = pd.get_dummies(df2, columns = category_list) 
df3

student_count  awards_per_value  awards_per_state_value  \
0          0.077136          0.212062                0.172680   
1          0.222567          0.342412                0.172680   
3          0.109244          0.342412                0.172680   
4          0.102608          0.161479                0.172680   
5          0.572686          0.291829                0.172680   
...             ...               ...                     ...   
3572       0.015751          0.262646                0.311856   
3639       0.076629          0.328794                0.304124   
3651       0.000371          0.291829                0.226804   
3717       0.008139          0.184825                0.275773   
3794       0.000800          0.258755                0.265464   

      awards_per_natl_value  exp_award_value  exp_award_state_value  \
0                  0.531915         0.175872               0.311746   
1                  0.531915         0.227992               0.311746   
3                  0.531915         0.107559               0.311746   
4                  0.531915         0.221081               0.311746   
5                  0.531915         0.125812               0.311746   
...                     ...              ...                    ...   
3572               0.638298         0.248117               0.288542   
3639               0.531915         0.077117               0.062524   
3651               0.638298         0.287266               0.500767   
3717               0.638298         0.308513               0.510470   
3794               0.638298         0.176930               0.616908   

      exp_award_natl_value    ft_pct  med_sat_value  aid_value  \
0                 0.474983  0.922597       0.180876   0.147490   
1                 0.474983  0.659176       0.552995   0.121398   
3                 0.474983  0.680400       0.592166   0.135236   
4                 0.474983  0.887640       0.188940   0.150312   
5                 0.474983  0.877653       0.581797   0.227894   
...                    ...       ...            ...        ...   
3572              1.000000  0.977528       0.465438   0.353476   
3639              0.474983  0.575531       0.465438   0.094589   
3651              1.000000  0.915106       0.465438   0.093425   
3717              1.000000  0.943820       0.569124   0.458313   
3794              1.000000  0.901373       0.464286   0.203263   

      grad_150_value  level_2-year  level_4-year  control_Private for-profit  \
0           0.295132         False          True                       False   
1           0.542596         False          True                       False   
3           0.490872         False          True                       False   
4           0.255578         False          True                       False   
5           0.676471         False          True                       False   
...              ...           ...           ...                         ...   
3572        0.585193         False          True                       False   
3639        0.433063         False          True                       False   
3651        0.000000         False          True                       False   
3717        0.622718         False          True                       False   
3794        0.522312         False          True                       False   

      control_Private not-for-profit  control_Public  
0                              False            True  
1                              False            True  
3                              False            True  
4                              False            True  
5                              False            True  
...                              ...             ...  
3572                            True           False  
3639                           False            True  
3651                            True           False  
3717                            True           False  
3794         

In [92]:
df3 = df3.drop(['student_count'],axis=1)

In [93]:
df3.grad_150_value.describe()

count    1319.000000
mean        0.553067
std         0.183258
min         0.000000
25%         0.428499
50%         0.543611
75%         0.673935
max         1.000000
Name: grad_150_value, dtype: float64

In [94]:
df3['grad_150'] = pd.cut(df3.grad_150_value, bins = [-1,0.67,1], labels =[0,1])
prevalence = df3.grad_150.value_counts()[1]/len(df3.grad_150)
prevalence 

np.float64(0.25701288855193327)

In [95]:
Train, Test = train_test_split(df3,  train_size = .8, stratify = df3.grad_150) 
Tune, Test = train_test_split(Test,  train_size = .5, stratify= Test.grad_150)
print(Train.shape)
print(Test.shape)
print(Tune.shape)

(1055, 16)
(132, 16)
(132, 16)


##### Job Placement Dataset

##### Independent Business Metric: Salary

##### Data Preparation

In [111]:
url = "https://raw.githubusercontent.com/DG1606/CMS-R-2020/master/Placement_Data_Full_Class.csv"

ds = pd.read_csv(url)

ds.head()

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0      1      M  67.00   Others  91.00   Others  Commerce     58.00   
1      2      M  79.33  Central  78.33   Others   Science     77.48   
2      3      M  65.00  Central  68.00  Central      Arts     64.00   
3      4      M  56.00  Central  52.00  Central   Science     52.00   
4      5      M  85.80  Central  73.60  Central  Commerce     73.30   

    degree_t workex  etest_p specialisation  mba_p      status    salary  
0   Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1   Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2  Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3   Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4  Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0

In [112]:
ds=ds.dropna()

In [113]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
Index: 148 entries, 0 to 213
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           148 non-null    int64  
 1   gender          148 non-null    object 
 2   ssc_p           148 non-null    float64
 3   ssc_b           148 non-null    object 
 4   hsc_p           148 non-null    float64
 5   hsc_b           148 non-null    object 
 6   hsc_s           148 non-null    object 
 7   degree_p        148 non-null    float64
 8   degree_t        148 non-null    object 
 9   workex          148 non-null    object 
 10  etest_p         148 non-null    float64
 11  specialisation  148 non-null    object 
 12  mba_p           148 non-null    float64
 13  status          148 non-null    object 
 14  salary          148 non-null    float64
dtypes: float64(6), int64(1), object(8)
memory usage: 18.5+ KB


In [114]:
ds[['ssc_b','hsc_b','hsc_s','degree_t','workex','specialisation']] = ds[['ssc_b','hsc_b','hsc_s','degree_t','workex','specialisation']].astype('category')

print(ds.dtypes)

sl_no                int64
gender              object
ssc_p              float64
ssc_b             category
hsc_p              float64
hsc_b             category
hsc_s             category
degree_p           float64
degree_t          category
workex            category
etest_p            float64
specialisation    category
mba_p              float64
status              object
salary             float64
dtype: object


In [115]:
abc = list(ds.select_dtypes('number'))

ds[abc] = MinMaxScaler().fit_transform(ds[abc])
ds

sl_no gender     ssc_p    ssc_b     hsc_p    hsc_b     hsc_s  \
0    0.000000      M  0.445545   Others  0.857051   Others  Commerce   
1    0.004695      M  0.750743  Central  0.586729   Others   Science   
2    0.009390      M  0.396040  Central  0.366332  Central      Arts   
4    0.018779      M  0.910891  Central  0.485812  Central  Commerce   
7    0.032864      M  0.816832  Central  0.280990  Central   Science   
..        ...    ...       ...      ...       ...      ...       ...   
209  0.981221      M  0.321782  Central  0.451675  Central  Commerce   
210  0.985915      M  0.782178   Others  0.665031   Others  Commerce   
211  0.990610      M  0.222772   Others  0.195648   Others   Science   
212  0.995305      M  0.445545   Others  0.344997   Others  Commerce   
213  1.000000      F  0.618812   Others  0.323661   Others  Commerce   

     degree_p   degree_t workex   etest_p specialisation     mba_p  status  \
0    0.057143   Sci&Tech     No  0.104167         Mkt&HR  0.251666  Placed   
1    0.613714   Sci&Tech    Yes  0.760417        Mkt&Fin  0.544884  Placed   
2    0.228571  Comm&Mgmt     No  0.520833        Mkt&Fin  0.212466  Placed   
4    0.494286  Comm&Mgmt     No  0.975000        Mkt&Fin  0.122305  Placed   
7    0.285714   Sci&Tech    Yes  0.354167        Mkt&Fin  0.382595  Placed   
..        ...        ...    ...       ...            ...       ...     ...   
209  0.257143  Comm&Mgmt     No  0.354167        Mkt&Fin  0.161113  Placed   
210  0.617143  Comm&Mgmt     No  0.854167        Mkt&Fin  0.866719  Placed   
211  0.457143   Sci&Tech     No  0.500000        Mkt&Fin  0.048608  Placed   
212  0.485714  Comm&Mgmt    Yes  0.187500        Mkt&Fin  0.679733  Placed   
213  0.057143  Comm&Mgmt     No  0.416667         Mkt&HR  0.307722  Placed   

       salary  
0    0.094595  
1    0.000000  
2    0.067568  
4    0.304054  
7    0.070270  
..        ...  
209  0.021622  
210  0.270270  
211  0.101351  
212  0.128378  
213  0.005405  

[148 rows x 15 columns]

In [116]:
category_list = list(ds.select_dtypes('category')) 
category_list
ds = pd.get_dummies(ds, columns = category_list) 
ds

sl_no gender     ssc_p     hsc_p  degree_p   etest_p     mba_p  \
0    0.000000      M  0.445545  0.857051  0.057143  0.104167  0.251666   
1    0.004695      M  0.750743  0.586729  0.613714  0.760417  0.544884   
2    0.009390      M  0.396040  0.366332  0.228571  0.520833  0.212466   
4    0.018779      M  0.910891  0.485812  0.494286  0.975000  0.122305   
7    0.032864      M  0.816832  0.280990  0.285714  0.354167  0.382595   
..        ...    ...       ...       ...       ...       ...       ...   
209  0.981221      M  0.321782  0.451675  0.257143  0.354167  0.161113   
210  0.985915      M  0.782178  0.665031  0.617143  0.854167  0.866719   
211  0.990610      M  0.222772  0.195648  0.457143  0.500000  0.048608   
212  0.995305      M  0.445545  0.344997  0.485714  0.187500  0.679733   
213  1.000000      F  0.618812  0.323661  0.057143  0.416667  0.307722   

     status    salary  ssc_b_Central  ...  hsc_s_Arts  hsc_s_Commerce  \
0    Placed  0.094595          False  ...       False            True   
1    Placed  0.000000           True  ...       False           False   
2    Placed  0.067568           True  ...        True           False   
4    Placed  0.304054           True  ...       False            True   
7    Placed  0.070270           True  ...       False           False   
..      ...       ...            ...  ...         ...             ...   
209  Placed  0.021622           True  ...       False            True   
210  Placed  0.270270          False  ...       False            True   
211  Placed  0.101351          False  ...       False           False   
212  Placed  0.128378          False  ...       False            True   
213  Placed  0.005405          False  ...       False            True   

     hsc_s_Science  degree_t_Comm&Mgmt  degree_t_Others  degree_t_Sci&Tech  \
0            False               False            False               True   
1             True               False            False               True   
2            False                True            False              False   
4            False                True            False              False   
7             True               False            False               True   
..             ...                 ...              ...                ...   
209          False                True            False              False   
210          False                True            False              False   
211           True               False            False               True   
212          False                True            False              False   
213          False                True            False              False   

     workex_No  workex_Yes  specialisation_Mkt&Fin  specialisation_Mkt&HR  
0         True       False                   False                   True  
1        False        True                    True                  False  
2         True       False                    True                  False  
4         True       False                    True                  False  
7        False        True                    True                  False  
..         ...         ...                     ...                    ...  
209       True       False                    True                  False  
210       True       False                    True                  False  
211       True       False                    True                  False  
212      False        True                    True                  False  
213       True       False                   False                   True  

[148 rows x 23 columns]

In [117]:
ds = ds.drop(['sl_no','gender','status'],axis=1)

In [118]:
ds.salary.describe()

count    148.000000
mean       0.119805
std        0.126294
min        0.000000
25%        0.054054
50%        0.087838
75%        0.135135
max        1.000000
Name: salary, dtype: float64

In [120]:
ds['salaryf'] = pd.cut(ds.salary, bins = [-1,0.135,1], labels =[0,1])
prevalence = ds.salaryf.value_counts()[1]/len(ds.salaryf)
prevalence 

np.float64(0.33783783783783783)

In [121]:
Train, Test = train_test_split(ds,  train_size = .8, stratify = ds.salaryf) 
Tune, Test = train_test_split(Test,  train_size = .5, stratify= Test.salaryf)
print(Train.shape)
print(Test.shape)
print(Tune.shape)

(118, 21)
(15, 21)
(15, 21)


## Step 3 

For the first dataset, I emphasized the important of student performance when it comes to determining a college's graduation rate. However, my concern with that dataset is that most of the information is the demographics pf the student population along with the SAT score. There was not even information on the academic performance, such as GPA, class rank average, and ACT score in fact. There are a lot more features to be concerned about in order to yield an accurate result and answer the question. For the second dataset, I chose salary as the target variable. But after I did the test and training split, I realized there wasn't a lot of data for the testing. There are only 148 entries for salary and the 80-20 split was not beneficial for the model, which can be another concern when ensuring the accuracy of the model. 